### PR TITLE
feat(sdk): supporting bring your own container for arbitrary input and outputs 

### PR DESCRIPTION
**Pull request by zichuan-scott-xu**
**Description of your changes:**
The first 16 commits handling functions with no inputs are already reviewed and approved in https://github.com/kubeflow/pipelines/pull/7973. 

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

File: sdk/python/kfp/components/container_component_decorator.py
Comment (connor-mccarthy) [inline sdk/python/kfp/components/container_component_decorator.py:50]: Can you take a look at the `@kfp.dsl.component` docstring and make sure this follows the same style? This will become a part of our API reference docs when merged with master.

You can inspect the docs by running:

```sh
cd docs
pip install -r requirements.txt
make clean html
open _build/html/index.html
```

File: sdk/python/kfp/dsl/__init__.py
Comment (connor-mccarthy) [inline sdk/python/kfp/dsl/__init__.py:46]: Can you add `container_component` and `ComponentSpec` to `__all__`? This explicitly declares these objects as belonging to the KFP SDK public API and ensures they are documented in our API reference docs.

This is required after the recent merge of master into this branch, but is very easy to miss since it didn't result in any conflicts.

File: sdk/python/kfp/dsl/__init__.py
Comment (connor-mccarthy) [inline sdk/python/kfp/dsl/__init__.py:50]: As with `container_component`, `ContainerSpec` will now be included in our API reference docs once this branch is merged with master.

Can you refresh the docstrings of `ContainerSpec` for this? See the docstrings of [PipelineTaskFinalStatus](https://github.com/kubeflow/pipelines/blob/b10440d6a8a3a67f0b6d8ddeed6d8befbb98cdf5/sdk/python/kfp/components/task_final_status.py#L20-L55) for an example of how to document dataclasses in a way that works well with our documentation generator.

A few things I noticed are:
* put attribute docstrings below attributes
* remove the "The" prefix on all descriptions, capitalize first letter of first word of description
* remove `(optional)` (type annotations convey this already)
* perhaps state that this is object is only meant to be used by pipeline authors when constructing an `@container_component`

**Review by connor-mccarthy — CHANGES_REQUESTED** [review]
This looks great! Just a few suggestions before we merge with master.

File: sdk/python/kfp/components/component_factory.py
Comment (zichuan-scott-xu) [reply sdk/python/kfp/components/component_factory.py:470]: Thanks for this comment! In theory it could, but that would require a refactoring of the extract_component_interface function. The main difference is that we don't parse the return type annotation for container_component early in the function. It would be possible to refactor this function by separating out the logic of return type annotation and determine whether we need it based on the type of the component. Do you think this refactoring is reasonable?

File: samples/v2/pipeline_container_no_input.py
Comment (chensun) [inline samples/v2/pipeline_container_no_input.py:38]: nit: Recommended file extension is now YAML. You'll get some warning when compiling to JSON.

File: samples/v2/two_step_pipeline_containerized.py
Comment (chensun) [inline samples/v2/two_step_pipeline_containerized.py:40]: nit: how about test `.path` here? It suppose to resolved to a local path that you can use just `cat` to show its content.

File: samples/v2/two_step_pipeline_containerized.py
Comment (chensun) [inline samples/v2/two_step_pipeline_containerized.py:43]: I think this one should result a compile error. Pipeline name cannot contain underscore: https://github.com/kubeflow/pipelines/blob/2a48096dc3cda7df45f15a7e029adda117064cc3/sdk/python/kfp/compiler/pipeline_spec_builder.py#L1256

File: samples/v2/two_step_pipeline_containerized.py
Comment (chensun) [inline samples/v2/two_step_pipeline_containerized.py:55]: ditto: yaml instead of json

File: samples/v2/two_step_pipeline_containerized_test.py
Comment (chensun) [inline samples/v2/two_step_pipeline_containerized_test.py:72]: Is this still the case?

Comment (zichuan-scott-xu) [reply samples/v2/two_step_pipeline_containerized_test.py:72]: I guess so since this sample test passes. Same comment is left in `pipeline_with_importer_test.py`.

File: samples/v2/two_step_pipeline_containerized.py
Comment (zichuan-scott-xu) [reply samples/v2/two_step_pipeline_containerized.py:40]: Sure! I will change this to path so that both `uri` and `path` are covered.

File: samples/v2/two_step_pipeline_containerized_test.py
Comment (connor-mccarthy) [reply samples/v2/two_step_pipeline_containerized_test.py:72]: Looks like this comment has gone stale -- the artifact type written to yaml is `system.Dataset`. @zichuan-scott-xu, can you remove?

**Review by connor-mccarthy — CHANGES_REQUESTED** [review]
Thanks for addressing the previous comments, Scott. Just a few more things I noticed from another review.

File: samples/v2/two_step_pipeline_containerized_test.py
Comment (zichuan-scott-xu) [reply samples/v2/two_step_pipeline_containerized_test.py:72]: Thanks for the reminder! I will remove the comments.

File: sdk/python/kfp/components/component_factory.py
Comment (connor-mccarthy) [inline sdk/python/kfp/components/component_factory.py:65]: Missing f in f-string. Consider using `self.assertRaisesRegex` in your test code to check that this string is constructed correctly.

File: sdk/python/kfp/components/component_factory.py
Comment (connor-mccarthy) [inline sdk/python/kfp/components/component_factory.py:324]: nit: Consider `omitted` instead of `ignored`. Ignored sounds like the KFP SDK ignores it, but we're actually talking about the author here.

File: sdk/python/kfp/compiler/test_data/components/container_with_artifact_output.py
Comment (chensun) [inline sdk/python/kfp/compiler/test_data/components/container_with_artifact_output.py:34]: What if user just puts `model` here? Do we throw an error or return the whole artifact json? Also what if users access some arbitrary valid or invalid property/method here? Do we throw errors?

**Review by chensun — COMMENTED** [review]
/lgtm

Great work, @zichuan-scott-xu !

I left one question, but it doesn't necessarily block this PR from merging.

File: sdk/python/kfp/compiler/test_data/components/container_with_artifact_output.py
Comment (zichuan-scott-xu) [reply sdk/python/kfp/compiler/test_data/components/container_with_artifact_output.py:34]: Thanks for the reminder! The current design is to throw error in both cases and I will add the error handling logic before merging the PR.

**Review by connor-mccarthy — CHANGES_REQUESTED** [review]
@zichuan-scott-xu, can you add a test to validate that the exception in [bb282dc](https://github.com/kubeflow/pipelines/pull/8066/commits/bb282dc5609f741310294cc28495551de1058091) is raised correctly?

**Review by connor-mccarthy — APPROVED** [review]
Great work, Scott!
/lgtm

cc: @chensun

**Comment by chensun** [comment]
/approve

Great work, @zichuan-scott-xu !

**Pull request merged**
